### PR TITLE
feat(HQM-377): in app sushi swap

### PR DESCRIPTION
**Pull request by iGroza**
Fixes #

## Proposed Changes

- HQM-377 in app sushi swap

## Checklist

- [x] I have read the rules of [**contribution**](https://github.com/haqq-network/haqq-wallet/blob/main/docs/contribution.md#important-steps).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -72,6 +72,8 @@
     "@react-navigation/stack": "6.3.17",
     "@sentry/react-native": "5.7.1",
     "@tradle/react-native-http": "2.0.1",
+    "@uniswap/sdk": "^3.0.3",
```
Comment (ragozin-nikita) [inline package.json:75]: Please fix the version number
```suggestion
    "@uniswap/sdk": "3.0.3",
```

File: src/components/swap/swap.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    isApproveInProgress,
+    isSwapInProgress,
+    estimate,
+    onPressChangeTokenIn,
+    onPressChangeTokenOut,
+    onPressApprove,
+    onPressSwap,
+  }: SwapProps) => {
+    const needApproveTokenIn = estimateData?.need_approve;
+    const isLoading = isEstimating;
```
Comment (ragozin-nikita) [inline src/components/swap/swap.tsx:64]: https://eslint.org/blog/2024/04/eslint-v9.0.0-released/#new-rule%3A-no-useless-assignment

😄

File: src/components/swap/swap.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            onPress={onPressChangeTokenOut}
+          />
+        </View>
+
+        <Spacer height={10} />
+
+        {!!estimateData && (
+          <View>
+            <Text variant={TextVariant.t11}>
+              Provider fee: {new Balance(estimateData.fee.amount).toFloat()}
```
Comment (ragozin-nikita) [inline src/components/swap/swap.tsx:133]: i18n

File: src/screens/SwapStack/swap-finish-screen.tsx [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+import {observer} from 'mobx-react';
+
+import {SwapFinish} from '@app/components/swap';
+
+export const SwapFinishScreen = observer(() => {
```
Comment (ragozin-nikita) [inline src/screens/SwapStack/swap-finish-screen.tsx:7]: You should use observer only in the component, that need to react to something. If there is not references to mobx actions/observables, then you can stay with memo

File: src/screens/SwapStack/swap-screen.tsx [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+      const amountOut =
+        parseInt(response.amount_out, 16) / Math.pow(10, tokenOut?.decimals!);
+      amountsOut.setAmount(amountOut.toString());
+    } catch (err) {
+      logger.captureException(err, 'estimate');
+    } finally {
+      setIsEstimating(false);
+    }
+  }, [
+    params,
```
Comment (ragozin-nikita) [inline src/screens/SwapStack/swap-screen.tsx:94]: I think you can skip memoization for this function, there is too many arguments for this

File: src/services/indexer/indexer.ts [hunk 6]
```diff
@@ -245,4 +252,58 @@ export class Indexer {
...
+      throw err;
+    }
+  }
+
+  async sushiPoolEstimate({
+    amount,
+    sender,
+    token_in,
+    token_out,
+    currency_id = Currencies.currency?.id,
```
Comment (ragozin-nikita) [inline src/services/indexer/indexer.ts:282]: sushiPoolEstimate is not an action and Indexer is not an observer, so `Currencies.currency` may have not relevant value - keep it in mind please

File: src/widgets/index.tsx [hunk 7]
```diff
@@ -29,32 +30,44 @@ type IWidgetMap = {
...
       deep?: boolean;
+      id?: string;
     },
   ) => ReactNode;
 };
 
 const WidgetMap: IWidgetMap = {
   Transactions: params => (
-    <TransactionsWidgetWrapper key={generateUUID()} {...params} />
+    <TransactionsWidgetWrapper key={params.id ?? generateUUID()} {...params} />
```
Comment (ragozin-nikita) [inline src/widgets/index.tsx:40]: Lets convert it to
```suggestion
    <TransactionsWidgetWrapper key={generateUUID(params)} {...params} />
```

And we will have incapsulated logic about UUID generation in `generateUUID` function

File: src/widgets/swap-widget/index.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    const address = await awaitForWallet({
+      wallets: Wallet.getAll(),
+      title: I18N.selectAccount,
+    });
+    navigation.navigate(HomeStackRoutes.Swap, {address});
+  }, [navigation]);
+
+  return (
+    <ShadowCard style={styles.wrapper} onPress={onPress}>
+      <View style={styles.image}>
```
Comment (ragozin-nikita) [inline src/widgets/swap-widget/index.tsx:29]: Height and width should be applied to image iteself -- there is no need to have extra wrapper

File: src/components/swap/swap.tsx (hunk 2 above)
Comment (ragozin-nikita) [reply src/components/swap/swap.tsx:64]: You can rename variable in props, if you don't like

File: src/screens/SwapStack/swap-screen.tsx (hunk 5 above)
Comment (iGroza) [reply src/screens/SwapStack/swap-screen.tsx:94]: I prefer to control the dependencies of useMemo/useCallback/useEffect for more precise behavior tuning. Moreover, this component has many other variables that didn’t make it into the dependencies.

**Review by ragozin-nikita — APPROVED** [review]

**Pull request merged**